core: introduce PickSubchannelArgs

lukaszx0 opened the pull request:
This will allow adding new arguments passed to pickSubchannel without significant API changes in core LB API in the future.

Addresses #2737

cc @ejona86 @zhangkun83

zhangkun83 commented on core/src/main/java/io/grpc/LoadBalancer2.java:
```diff
@@ -317,6 +322,77 @@ public String toString() {
   /**
+   * Provides arguments for a {@link SubchannelPicker#pickSubchannel(PickSubchannelArgs)}.
+   */
+  public abstract static class PickSubchannelArgs {
```
Since LB impls will provide this object, we should make it convenient to create, probably a final class with a builder.

lukaszx0 replied:
Good call.

zhangkun83 replied:
Oh sorry, I must have gone out of my mind. The args are *consumed* by LB impls, so an abstract class as described in #2737 is most appropriate.

lukaszx0 replied:
I'm confused. The final immutable class still makes sense to me in this context, no? Do you anticipate passing different types of `PickSubchannelArgs` to `pickSubchannel` internally? If so when and why?

lukaszx0 replied:
I guess the only change needed would be to either add interface in `LoadBalancer2` or promote `PickSubchannelArgs` from internal to `io.grpc`. Probably latter, since this will let implementors re-use it in tests.

zhangkun83 replied:
`PickSubchannelArgs` is public API, thus it should be in `io.grpc` (preferably as inner of `LoadBalancer2`) instead of `io.grpc.internal`.

Builder pattern exposes the construction of `PickSubchannelArgs` as public API, which is not needed by LB impls because they are only consumers of `PickSubchannelArgs`. The builder API has to define required fields (headers) and enforce it, but there isn't an ideal way to do it (there was a [previous discussion](https://github.com/grpc/grpc-java/pull/2667#discussion_r98262481) on a similar issue). It's better to avoid exposing such a builder unless it's necessary.

lukaszx0 replied:
How would you unit test your custom impl of `Picker` without access to builder? Exposing only abstract class would require you to implement it which effectively will be just copy of what's used internally because this class is just a simple container.

Anyway, unclear at this point if you think we should revert it back to `abstract` or make it `public final` (as it is now) ?

zhangkun83 replied:
I am thinking about reverting back to `abstract`. The channel would have its own private implementation which will not have a builder, but just a constructor. Unit tests would extend the abstract class to make fake values. Unfortunately Mockito would be discouraged because any unstubbed methods (e.g., new fields) will return null.

lukaszx0 replied:
That puts a bit of a burden on implementors when writing tests. Not a strong opinion but I slightly favour builder though I understand your point of view and I'm fine reverting that back. Let me know what do you think works better for you.

zhangkun83 replied:
I agree with you it's a bit inconvenient for test writers. Because it appears to me both approaches have about the same amount of problems, I would opt for a simpler API, which is the abstract class.

zhangkun83 replied:
And then the class would be small enough to fit in `LoadBalancer2`.

lukaszx0 replied:
Reverted back to abstract. PTAL.

lukaszx0 commented:
@zhangkun83 regarding `methodDescriptor`, I'd argue that it's very useful to have it and there's already a lot of use cases for it, first that immediately comes to mind - ability to have code recording metrics inside picker, based on method's name. I think we should provide it, it'll be very helpful for implementors, myself included. Don't have very strong opinion, can remove it if you're not convinced.

Addressed other comments, PTAL.

lukaszx0 commented on core/src/main/java/io/grpc/LoadBalancer2.java:
```diff
@@ -170,13 +171,17 @@ public abstract void handleSubchannelState(
+     * @param args the pick arguments
      */
-    public abstract PickResult pickSubchannel(Attributes affinity, Metadata headers);
+    public abstract PickResult pickSubchannel(PickSubchannelArgs args);
```
AH yeah, good point. Thanks!

lukaszx0 commented on core/src/main/java/io/grpc/LoadBalancer2.java:
```diff
@@ -317,6 +322,77 @@ public String toString() {
+    /**
+     * Call metadata.
+     */
+    public abstract Metadata getMetadata();
```
👍

lukaszx0 commented:
> You may want to wait until I merge #2743 before making further changes.

It's OK, I've addressed feedback and will rebase after your migration lands. Thanks!

lukaszx0 commented on core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  @Override
+  public String toString() {
```
Done.

lukaszx0 commented on core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java:
```diff
@@ -0,0 +1,102 @@
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(callOptions, headers, method);
+  }
```
Good call, though it convinces me a bit that maybe using builder in this case would've been more convenient

ejona86 replied:
Ewww... Um, the throwing problem sort of makes me think we shouldn't implement `toString()` in the abstract class either. Granted, we could call a method within try-catch in order to handle an exception, but it is still pretty icky.

The only reason this is an abstract class instead of interface is to allow adding new methods in the future. So in that way, I don't mind requiring toString to be implemented in each extending class.

lukaszx0 commented on core/src/main/java/io/grpc/LoadBalancer2.java:
```diff
@@ -317,6 +325,77 @@ public String toString() {
+    /**
+     * Call affinity.
+     */
+    public Attributes getAffinity() {
```
Done.

lukaszx0 commented on core/src/main/java/io/grpc/LoadBalancer2.java:
```diff
@@ -170,13 +171,20 @@ public abstract void handleSubchannelState(
+     * @param args the pick arguments
      */
-    public abstract PickResult pickSubchannel(Attributes affinity, Metadata headers);
+    // TODO(lukasz) make it public once deprecated overload will be removed.
```
Doh, I meant _abstract_ 🤦‍♂️

lukaszx0 commented on core/src/main/java/io/grpc/LoadBalancer2.java:
```diff
@@ -170,13 +171,20 @@ public abstract void handleSubchannelState(
   public abstract static class SubchannelPicker {
+    @Deprecated
+    public PickResult pickSubchannel(Attributes affinity, Metadata headers) {
+      return pickSubchannel(new DeprecatedPickSubchannelArgs(affinity, headers));
```
Ah, good point. Removed.

lukaszx0 commented on core/src/main/java/io/grpc/LoadBalancer2.java:
```diff
@@ -170,13 +171,20 @@ public abstract void handleSubchannelState(
    */
   @ThreadSafe
   public abstract static class SubchannelPicker {
+    @Deprecated
```
Done.

zhangkun83 commented on core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java:
```diff
@@ -0,0 +1,70 @@
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+// TODO(lukaszx0) make this class pkg private after removing deprecated API in PickSubchannelArgs
```
It is already so.

lukaszx0 replied:
Ooops, leftover. Thanks!

zhangkun83 commented on core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java:
```diff
@@ -71,6 +72,7 @@
 
   private static final Attributes.Key<String> FOO = Attributes.Key.of("foo");
   private Attributes affinity = Attributes.newBuilder().set(FOO, "bar").build();
```
Remove this?

lukaszx0 replied:
It is still used for `handleResolvedAddresses` calls.

zhangkun83 commented on core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java:
```diff
@@ -71,6 +72,7 @@
 
   private static final Attributes.Key<String> FOO = Attributes.Key.of("foo");
   private Attributes affinity = Attributes.newBuilder().set(FOO, "bar").build();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
Only if you verify at the end of the test that `EMPTY_ARGS` never had an interaction, it's fine to mock the args without any stubbing.

lukaszx0 replied:
Added `tearDown` with `verifyNoMoreInteractions`

zhangkun83 commented on core/src/test/java/io/grpc/util/RoundRobinLoadBalancer2Test.java:
```diff
@@ -87,6 +87,7 @@
   private Map<EquivalentAddressGroup, Subchannel> subchannels = Maps.newLinkedHashMap();
   private static final Attributes.Key<String> MAJOR_KEY = Attributes.Key.of("major-key");
   private Attributes affinity = Attributes.newBuilder().set(MAJOR_KEY, "I got the keys").build();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
dittos with `PickFirstBalancerFactory2`

lukaszx0 replied:
Added `tearDown` with `verifyNoMoreInteractions`

zhangkun83 commented on grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancer2Test.java:
```diff
@@ -157,7 +157,7 @@ public String parse(InputStream stream) {
   private final SerializingExecutor channelExecutor =
       new SerializingExecutor(MoreExecutors.directExecutor());
-  private final Metadata headers = new Metadata();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
dittos with PickFirstBalancerFactory2

lukaszx0 replied:
Added `tearDown` with `verifyNoMoreInteractions`

lukaszx0 commented:
I didn't realize that LBv2 transition will create such bad conflicts so I pretty much just did it over again - it was just easier to do so some copy-pasting from the old version.

@zhangkun83 PTAL and review again as a whole. Thanks!

zhangkun83 commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -318,7 +318,7 @@ void exitIdleMode() {
   // Must be run from channelExecutor
   private void cancelIdleTimer() {
-    if (idleModeTimerFuture != null) {
+    if (idleModeTimerFuture != null && idleModeTimer != null) {
```
Please don't change this line.
 1. This is not related to the subject of this PR.
 2. `idleModeTimer` should be synchronized with `idleModeTimerFuture` in terms of nullness. If one is null while the other is not, it must be a bug and we want to let it throw instead of hiding it.

lukaszx0 replied:
Damn! I didn't intend to include this, I was experimenting with some tooling and this slipped somehow. Sorry, reverted.

zhangkun83 commented on core/src/main/java/io/grpc/internal/PickSubchannelArgsImpl.java:
```diff
@@ -0,0 +1,69 @@
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+class PickSubchannelArgsImpl extends PickSubchannelArgs {
```
final

lukaszx0 replied:
Done.

zhangkun83 commented on core/src/main/java/io/grpc/LoadBalancer.java:
```diff
@@ -169,8 +170,75 @@ public abstract void handleSubchannelState(
+     * getter values.
+     */
+    @Override
+    public boolean equals(Object o) {
```
final

ejona86 replied:
Actually, let's not have equals/hashCode. If you're comparing them, Your Doing It Wrong™. That also prevents having the problem later where it is hard to figure out what to consider "equal".

lukaszx0 replied:
Done

zhangkun83 commented on core/src/main/java/io/grpc/LoadBalancer.java:
```diff
@@ -169,8 +170,75 @@ public abstract void handleSubchannelState(
+     * Returns a hash code for the subchannel args object.
+     */
+    @Override
+    public int hashCode() {
```
final

lukaszx0 replied:
Done

zhangkun83 commented on core/src/test/java/io/grpc/PickFirstLoadBalancerTest.java:
```diff
@@ -71,6 +73,7 @@
 
   private static final Attributes.Key<String> FOO = Attributes.Key.of("foo");
   private Attributes affinity = Attributes.newBuilder().set(FOO, "bar").build();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
We don't want to set an example for users and let them think they can mock `PickSubchannelArgs` whenever we want. We'd better put a comment here says because this LoadBalancer doesn't use any of the arg fields, as verified in `tearDown()`, it's safe to use a mock without any stubbing. `unusedArgs` seems to be a more accurate name. I would also make it non-static, so that if any tests violates this precondition, we will know exactly which they are.

lukaszx0 replied:
Good points. Renamed it to `mockArgs`, which stick nicely with other fields and the convention we have.

zhangkun83 commented on core/src/test/java/io/grpc/util/RoundRobinLoadBalancerTest.java:
```diff
@@ -87,6 +88,7 @@
   private Map<EquivalentAddressGroup, Subchannel> subchannels = Maps.newLinkedHashMap();
   private static final Attributes.Key<String> MAJOR_KEY = Attributes.Key.of("major-key");
   private Attributes affinity = Attributes.newBuilder().set(MAJOR_KEY, "I got the keys").build();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
ditto as in PickFirstLoadBalancerTest

zhangkun83 commented on grpclb/src/test/java/io/grpc/grpclb/GrpclbLoadBalancerTest.java:
```diff
@@ -157,7 +157,7 @@ public String parse(InputStream stream) {
   private final SerializingExecutor channelExecutor =
       new SerializingExecutor(MoreExecutors.directExecutor());
-  private final Metadata headers = new Metadata();
+  private static final PickSubchannelArgs EMPTY_ARGS = mock(PickSubchannelArgs.class);
```
ditto as in PickFirstLoadBalancerTest

lukaszx0 commented on core/src/main/java/io/grpc/internal/DelayedClientTransport.java:
```diff
@@ -155,7 +155,8 @@ public final ClientStream newStream(MethodDescriptor<?, ?> method, Metadata head
         while (true) {
-          PickResult pickResult = picker.pickSubchannel(callOptions.getAffinity(), headers);
+          PickResult pickResult = picker.pickSubchannel(
+              new PickSubchannelArgsImpl(method, headers, callOptions));
```
Tried to do a quick refactor, some tests were failing. Is it ok to address this in a follow up PR?

zhangkun83 replied:
Hmm, I didn't expect this to be more than a trivial change. What tests broke?

lukaszx0 replied:
It's possible that it is and I've made some stupid mistake, the `DelayedClientTransportTest` and one other tests were failing. I'll look into this again later today.

lukaszx0 commented on core/src/main/java/io/grpc/LoadBalancer.java:
```diff
@@ -169,8 +170,75 @@ public abstract void handleSubchannelState(
      *
      * @param affinity the affinity attributes provided via {@link CallOptions#withAffinity}
      * @param headers the headers container of the RPC. It can be mutated within this method.
+     * @deprecated use {@link #pickSubchannel(PickSubchannelArgs)}
```
Good call. Added javadocs.

lukaszx0 commented:
Since comments on diff collapsed, moving the unaddressed one from @ejona86 up for discussion here:

> Actually, let's not have equals/hashCode. If you're comparing them, Your Doing It Wrong™. That also prevents having the problem later where it is hard to figure out what to consider "equal".

I agree on the principle, but turns our our tests are relying on this (as will probably the ones implemented by users).

@ejona86 @zhangkun83 thoughts?

ejona86 commented:
> but turns our our tests are relying on this (as will probably the ones implemented by users)

Then the concrete implementation of PickSubchannelArgs used in the tests could implement the methods. The concrete implementation can easily avoid exceptions and API oddity.

lukaszx0 commented:
> Then the concrete implementation of PickSubchannelArgs used in the tests could implement the methods. The concrete implementation can easily avoid exceptions and API oddity.

Then it's fine to have them in `PickSubchannelArgsImpl` too, right?

lukaszx0 commented:
@ejona86 @zhangkun83 PTAL

zhangkun83 commented on core/src/main/java/io/grpc/internal/DelayedClientTransport.java:
```diff
@@ -347,26 +348,25 @@ public LogId getLogId() {
       this.statsTraceCtx = statsTraceCtx;
     }
 
+    public PickSubchannelArgs getArgs() {
```
This method is not necessary, because `args` is only used from the same file.

lukaszx0 replied:
Right, fixed.

zhangkun83 commented on core/src/main/java/io/grpc/internal/DelayedClientTransport.java:
```diff
@@ -290,24 +291,24 @@ final void reprocess(SubchannelPicker picker) {
+          public void run() {
+            stream.createRealStream(transport);
+          }
+        });
```
Please don't do any no-op reformatting.

zhangkun83 commented on core/src/test/java/io/grpc/internal/DelayedClientTransportTest.java:
```diff
@@ -479,89 +501,92 @@ public void reprocess_newStreamRacesWithReprocess() throws Exception {
+        // Will call pickSubchannel and wait on barrier
+        delayedTransport.newStream(method, headers, callOptions, statsTraceCtx);
+      }
+    };
```
ditto -- please refrain from any no-op reformatting.

zhangkun83 commented on core/src/test/java/io/grpc/internal/DelayedClientTransportTest.java:
```diff
@@ -479,89 +501,92 @@ public void reprocess_newStreamRacesWithReprocess() throws Exception {
+        // Will call pickSubchannel and wait on barrier
+        delayedTransport.newStream(method, headers2, callOptions, statsTraceCtx);
+      }
+    };
```
ditto

lukaszx0 commented:
@zhangkun83 reverted formatting that you pointed out and found it in few more places - sorry IJ did that to me ;) Should be good now, PTAL.

ejona86 commented:
I did some spot-checking, and this looks fine to me.

> Then it's fine to have them in `PickSubchannelArgsImpl` too, right?

Ideally, I'd say no, since applications could depend on the behavior. I think @zhangkun83 and I agree on that. However, some of the tests look very painful to fix if equals was dropped. So while disappointed, I'm okay with having equals/hashcode on the primary Impl because doing otherwise would be too onerous.

lukaszx0 commented:
@ejona86 thanks, sounds good! @zhangkun83 are we good to go with this?

zhangkun83 commented:
Please resolve the conflict.

lukaszx0 commented:
Oops, didn't notice there was a conflict. Fixed!
